### PR TITLE
LineageOnDemand Horizontal Pagination & Traversal Filters

### PR DESCRIPTION
**Pull request by BurakAltas**
## Change description

> META-3305 & META-3310 - LineageOnDemand horizontal pagination and traversal filters feature including Atlas UI updates

## Type of change
- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java [hunk 1]
```diff
@@ -645,8 +685,9 @@ private void addEdgeToResult(AtlasEdge edge, AtlasLineageInfo lineageInfo,
         }
     }
 
+
     private void addEdgeToResult(AtlasEdge edge, AtlasLineageInfo lineageInfo) throws AtlasBaseException {
-        if (!lineageContainsEdgeV2(lineageInfo, edge) && !lineageMaxNodeCountReached(lineageInfo.getRelations())) {
+        if (!lineageContainsEdge(lineageInfo, edge) && !lineageMaxNodeCountReached(lineageInfo.getRelations())) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java:690]: discuss

File: repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java [hunk 2]
```diff
@@ -402,6 +429,9 @@ private boolean incrementAndCheckIfRelationsLimitReached(AtlasEdge atlasEdge, bo
         String                     outGuid                   = AtlasGraphUtilsV2.getIdFromVertex(outVertex);
         LineageOnDemandConstraints outGuidLineageConstraints = getAndValidateLineageConstraintsByGuid(outGuid, lineageConstraintsMap);
 
+        boolean outVisitedFlag = ret.getRelationsOnDemand().containsKey(outGuid);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java:432]: discuss

File: repository/src/main/java/org/apache/atlas/discovery/LineageSearchProcessor.java [hunk 3]
```diff
@@ -79,6 +79,25 @@ protected Predicate constructInMemoryPredicate(AtlasTypeRegistry typeRegistry, S
         return ret;
     }
 
+    protected Predicate constructInMemoryPredicate(AtlasTypeRegistry typeRegistry, List<SearchParameters.FilterCriteria> filterCriteriaList) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/discovery/LineageSearchProcessor.java:82]: discuss

Comment (BurakAltas) [reply repository/src/main/java/org/apache/atlas/discovery/LineageSearchProcessor.java]: addressed

File: repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java (hunk 1 above)
Comment (BurakAltas) [reply repository/src/main/java/org/apache/atlas/discovery/EntityLineageService.java]: addressed

**Review by mehtaanshul — APPROVED** [review]

**Pull request merged**